### ISSUE TITLE
Restore backup, Crash on sql, On Smartphone or Emualtor

### ISSUE THREAD

**Issue by shaltayl** [signalapp/Signal-Android#9133]
- [x] I have searched open and closed issues for duplicates ( the nearest is #7863 )
- [x] I am submitting a bug report for existing functionality that does not work as intended
- [x] I have read https://github.com/signalapp/Signal-Android/wiki/Submitting-useful-bug-reports
- [x] This isn't a feature request or a discussion topic

----------------------------------------

### Bug description
I have created a Backup with my Phone before performs a Factory reset (bt wipe data with TWRP)
I try to restore my bakup on the same phone (OnePlus5) and same Os Lineage OS 16.
The application crash after have enter the pass verify number. 
### Steps to reproduce
- Backup signal messages
- save the backup on a PC
- do a wipe all data usgin TWRP
- reconfigure the Lineage OS 16
- try to reinstall the backup

For complete information, I do some mistake (but this is not the bug filled)
I have install signal (without launching) and after I have transfert the backup file using the command : adb push signal-xxx.backup /sdcard/Signal/Backups  and I have launched Signal.
That does not work (Signal doesn't see the Backup and crash) because my backup file became the  /sdcard/Signal/Backups file (and not a directory).....
So I have cleared cache, storage and uninstall Signal.
After that I have did a new try with  adb push signal-xxx.backup /sdcard/Signal/Backups with a already created /sdcard/Signal/Backups  directory.
Signal crash during recover backups. I enter the verification numbers and the application see my 3000+ message in the backup and crash one second later...

I have tested to add all permissions (using system application configuration) but this does not change anything.

 

**Actual result: Application cras 
**Expected result: recover the backuped messages

### Device info

**Device:** OnePlus 5 A5000
**Android version:** LineageOS 16.0 (20190721)
**Signal version: 4.49.10 

### Link to debug log

10-27 14:00:19.316  5518  6875 I RestoreBackupFragment: Starting backup restore.
10-27 14:00:19.321  5518  6682 D OpenGLRenderer: endAllActiveAnimators on 0x70fd7d0e00 (RippleDrawable) with handle 0x70fe449260
10-27 14:00:19.479   803  1372 W SurfaceFlinger: Attempting to set client state on removed layer: org.thoughtcrime.securesms/org.thoughtcrime.securesms.registration.RegistrationNavigationActivity#1
10-27 14:00:19.479   803  1372 W SurfaceFlinger: Attempting to set client state on removed layer: Dim Layer for - Task=47#0
10-27 14:00:19.479   803  1372 W SurfaceFlinger: Attempting to destroy on removed layer: org.thoughtcrime.securesms/org.thoughtcrime.securesms.registration.RegistrationNavigationActivity#1
10-27 14:00:19.479   803  1372 W SurfaceFlinger: Attempting to destroy on removed layer: Dim Layer for - Task=47#0
10-27 14:00:19.751  6601  6601 W /system/bin/adbd: type=1400 audit(0.0:124): avc: denied { create } for comm=7365727665722076736F636B scontext=u:r:adbd:s0 tcontext=u:r:adbd:s0 tclass=socket permissive=0
10-27 14:00:19.911  5518  6875 V Cursor  : Filling cursor window with start position:0 required position:0
10-27 14:00:19.927  5518  6875 I FullBackupImporter: Ignoring import for statement: CREATE TABLE sqlite_sequence(name,seq)
10-27 14:00:20.543  3807  3906 W OpaSSRequestSender: Failed to get ZeroStatResponse from AssistantClientSyncWorker with error java.util.concurrent.TimeoutException
10-27 14:00:20.751  6601  6601 W /system/bin/adbd: type=1400 audit(0.0:125): avc: denied { create } for comm=7365727665722076736F636B scontext=u:r:adbd:s0 tcontext=u:r:adbd:s0 tclass=socket permissive=0
--------- beginning of crash
10-27 14:00:21.211  5518  6875 E AndroidRuntime: FATAL EXCEPTION: AsyncTask #6
10-27 14:00:21.211  5518  6875 E AndroidRuntime: Process: org.thoughtcrime.securesms, PID: 5518
10-27 14:00:21.211  5518  6875 E AndroidRuntime: java.lang.RuntimeException: An error occurred while executing doInBackground()
10-27 14:00:21.211  5518  6875 E AndroidRuntime:        at android.os.AsyncTask$3.done(AsyncTask.java:354)
10-27 14:00:21.211  5518  6875 E AndroidRuntime:        at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
10-27 14:00:21.211  5518  6875 E AndroidRuntime:        at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
10-27 14:00:21.211  5518  6875 E AndroidRuntime:        at java.util.concurrent.FutureTask.run(FutureTask.java:271)
10-27 14:00:21.211  5518  6875 E AndroidRuntime:        at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:245)
10-27 14:00:21.211  5518  6875 E AndroidRuntime:        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
10-27 14:00:21.211  5518  6875 E AndroidRuntime:        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
10-27 14:00:21.211  5518  6875 E AndroidRuntime:        at java.lang.Thread.run(Thread.java:764)
10-27 14:00:21.211  5518  6875 E AndroidRuntime: Caused by: net.sqlcipher.database.SQLiteException: no such table: recipient: , while compiling: SELECT _id FROM recipient WHERE phone = ?
10-27 14:00:21.211  5518  6875 E AndroidRuntime:        at net.sqlcipher.database.SQLiteCompiledSql.native_compile(Native Method)
10-27 14:00:21.211  5518  6875 E AndroidRuntime:        at net.sqlcipher.database.SQLiteCompiledSql.compile(SQLiteCompiledSql.java:3)
10-27 14:00:21.211  5518  6875 E AndroidRuntime:        at net.sqlcipher.database.SQLiteCompiledSql.<init>(SQLiteCompiledSql.java:12)
10-27 14:00:21.211  5518  6875 E AndroidRuntime:        at net.sqlcipher.database.SQLiteProgram.<init>(SQLiteProgram.java:18)
10-27 14:00:21.211  5518  6875 E AndroidRuntime:        at net.sqlcipher.database.SQLiteQuery.<init>(SQLiteQuery.java:1)
10-27 14:00:21.211  5518  6875 E AndroidRuntime:        at net.sqlcipher.database.SQLiteDirectCursorDriver.query(SQLiteDirectCursorDriver.java:8)
10-27 14:00:21.211  5518  6875 E AndroidRuntime:        at net.sqlcipher.database.SQLiteDatabase.rawQueryWithFactory(SQLiteDatabase.java:5)
10-27 14:00:21.211  5518  6875 E AndroidRuntime:        at net.sqlcipher.database.SQLiteDatabase.queryWithFactory(SQLiteDatabase.java:4)
10-27 14:00:21.211  5518  6875 E AndroidRuntime:        at net.sqlcipher.database.SQLiteDatabase.query(SQLiteDatabase.java:1)
10-27 14:00:21.211  5518  6875 E AndroidRuntime:        at net.sqlcipher.database.SQLiteDatabase.query(SQLiteDatabase.java:2)
10-27 14:00:21.211  5518  6875 E AndroidRuntime:        at org.thoughtcrime.securesms.database.RecipientDatabase.getByColumn(RecipientDatabase.java:3)
10-27 14:00:21.211  5518  6875 E AndroidRuntime:        at org.thoughtcrime.securesms.database.RecipientDatabase.getOrInsertByColumn(RecipientDatabase.java:2)
10-27 14:00:21.211  5518  6875 E AndroidRuntime:        at org.thoughtcrime.securesms.database.RecipientDatabase.getOrInsertFromE164(RecipientDatabase.java:1)
10-27 14:00:21.211  5518  6875 E AndroidRuntime:        at org.thoughtcrime.securesms.recipients.Recipient.external(Recipient.java:8)
10-27 14:00:21.211  5518  6875 E AndroidRuntime:        at org.thoughtcrime.securesms.backup.FullBackupImporter.processAvatar(FullBackupImporter.java:2)
10-27 14:00:21.211  5518  6875 E AndroidRuntime:        at org.thoughtcrime.securesms.backup.FullBackupImporter.importFile(FullBackupImporter.java:11)
10-27 14:00:21.211  5518  6875 E AndroidRuntime:        at org.thoughtcrime.securesms.registration.fragments.RestoreBackupFragment$2.doInBackground(RestoreBackupFragment.java:7)
10-27 14:00:21.211  5518  6875 E AndroidRuntime:        at org.thoughtcrime.securesms.registration.fragments.RestoreBackupFragment$2.doInBackground(RestoreBackupFragment.java:1)
10-27 14:00:21.211  5518  6875 E AndroidRuntime:        at android.os.AsyncTask$2.call(AsyncTask.java:333)
10-27 14:00:21.211  5518  6875 E AndroidRuntime:        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
10-27 14:00:21.211  5518  6875 E AndroidRuntime:        ... 4 more
10-27 14:00:21.211  5518  6875 E UncaughtExceptionLogger: 
10-27 14:00:21.211  5518  6875 E UncaughtExceptionLogger: java.lang.RuntimeException: An error occurred while executing doInBackground()
10-27 14:00:21.211  5518  6875 E UncaughtExceptionLogger:       at android.os.AsyncTask$3.done(AsyncTask.java:354)
10-27 14:00:21.211  5518  6875 E UncaughtExceptionLogger:       at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
10-27 14:00:21.211  5518  6875 E UncaughtExceptionLogger:       at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
10-27 14:00:21.211  5518  6875 E UncaughtExceptionLogger:       at java.util.concurrent.FutureTask.run(FutureTask.java:271)
10-27 14:00:21.211  5518  6875 E UncaughtExceptionLogger:       at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:245)
10-27 14:00:21.211  5518  6875 E UncaughtExceptionLogger:       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
10-27 14:00:21.211  5518  6875 E UncaughtExceptionLogger:       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
10-27 14:00:21.211  5518  6875 E UncaughtExceptionLogger:       at java.lang.Thread.run(Thread.java:764)
10-27 14:00:21.211  5518  6875 E UncaughtExceptionLogger: Caused by: net.sqlcipher.database.SQLiteException: no such table: recipient: , while compiling: SELECT _id FROM recipient WHERE phone = ?
10-27 14:00:21.211  5518  6875 E UncaughtExceptionLogger:       at net.sqlcipher.database.SQLiteCompiledSql.native_compile(Native Method)
10-27 14:00:21.211  5518  6875 E UncaughtExceptionLogger:       at net.sqlcipher.database.SQLiteCompiledSql.compile(SQLiteCompiledSql.java:3)
10-27 14:00:21.211  5518  6875 E UncaughtExceptionLogger:       at net.sqlcipher.database.SQLiteCompiledSql.<init>(SQLiteCompiledSql.java:12)
10-27 14:00:21.211  5518  6875 E UncaughtExceptionLogger:       at net.sqlcipher.database.SQLiteProgram.<init>(SQLiteProgram.java:18)
10-27 14:00:21.211  5518  6875 E UncaughtExceptionLogger:       at net.sqlcipher.database.SQLiteQuery.<init>(SQLiteQuery.java:1)
10-27 14:00:21.211  5518  6875 E UncaughtExceptionLogger:       at net.sqlcipher.database.SQLiteDirectCursorDriver.query(SQLiteDirectCursorDriver.java:8)
10-27 14:00:21.211  5518  6875 E UncaughtExceptionLogger:       at net.sqlcipher.database.SQLiteDatabase.rawQueryWithFactory(SQLiteDatabase.java:5)
10-27 14:00:21.211  5518  6875 E UncaughtExceptionLogger:       at net.sqlcipher.database.SQLiteDatabase.queryWithFactory(SQLiteDatabase.java:4)
10-27 14:00:21.211  5518  6875 E UncaughtExceptionLogger:       at net.sqlcipher.database.SQLiteDatabase.query(SQLiteDatabase.java:1)
10-27 14:00:21.211  5518  6875 E UncaughtExceptionLogger:       at net.sqlcipher.database.SQLiteDatabase.query(SQLiteDatabase.java:2)
10-27 14:00:21.211  5518  6875 E UncaughtExceptionLogger:       at org.thoughtcrime.securesms.database.RecipientDatabase.getByColumn(RecipientDatabase.java:3)
10-27 14:00:21.211  5518  6875 E UncaughtExceptionLogger:       at org.thoughtcrime.securesms.database.RecipientDatabase.getOrInsertByColumn(RecipientDatabase.java:2)
10-27 14:00:21.211  5518  6875 E UncaughtExceptionLogger:       at org.thoughtcrime.securesms.database.RecipientDatabase.getOrInsertFromE164(RecipientDatabase.java:1)
10-27 14:00:21.211  5518  6875 E UncaughtExceptionLogger:       at org.thoughtcrime.securesms.recipients.Recipient.external(Recipient.java:8)
10-27 14:00:21.211  5518  6875 E UncaughtExceptionLogger:       at org.thoughtcrime.securesms.backup.FullBackupImporter.processAvatar(FullBackupImporter.java:2)
10-27 14:00:21.211  5518  6875 E UncaughtExceptionLogger:       at org.thoughtcrime.securesms.backup.FullBackupImporter.importFile(FullBackupImporter.java:11)
10-27 14:00:21.211  5518  6875 E UncaughtExceptionLogger:       at org.thoughtcrime.securesms.registration.fragments.RestoreBackupFragment$2.doInBackground(RestoreBackupFragment.java:7)
10-27 14:00:21.211  5518  6875 E UncaughtExceptionLogger:       at org.thoughtcrime.securesms.registration.fragments.RestoreBackupFragment$2.doInBackground(RestoreBackupFragment.java:1)
10-27 14:00:21.211  5518  6875 E UncaughtExceptionLogger:       at android.os.AsyncTask$2.call(AsyncTask.java:333)
10-27 14:00:21.211  5518  6875 E UncaughtExceptionLogger:       at java.util.concurrent.FutureTask.run(FutureTask.java:266)
10-27 14:00:21.211  5518  6875 E UncaughtExceptionLogger:       ... 4 more
10-27 14:00:21.221  1265  1916 W ActivityManager:   Force finishing activity org.thoughtcrime.securesms/.registration.RegistrationNavigationActivity

**Comment by shaltayl** [comment]
I have always my bug.
I have updated my twrp with encrypt support (last codework twrp version)
Wipe everything, even data, installed the last Lineage OS, update the firmware package to the last version .
I have still the problem
The logcat of different problem can be see here: 
https://pastebin.com/raw/jVEZj63q

The root of the problem is perhaps keystore, there is a Warning when Signal try to perform a "android.security.KeyStore.get" 

Thank

**Comment by shaltayl** [comment]
Hi,

Please, help me to recovered my backup, I have important thing inside. 
backup should be crutial and robust functionnality in Signal.

I perform two tests today (use signal-back application and use emulator).


I have tested the application signal-back, I can recovered images inside the backup, but I have not seen where are messages or keys.

```
signal-back/src/github.com/xeals/signal-back$ ./signal-back check  signal-2019-10-26-16-51-59.backup 
Password: 2019/11/01 22:37:16 Backup looks okay from here.
```


I have tested emulator Nexus-5x on the latest android-sudio and setup the 4.49.13 apk from the website.
When I try to recover the Backup in the emultor , I got the same error than with my OnePlus5 phone:

```
--------- beginning of crash
11-01 22:31:37.868  7988  8071 E AndroidRuntime: FATAL EXCEPTION: AsyncTask #2
11-01 22:31:37.868  7988  8071 E AndroidRuntime: Process: org.thoughtcrime.securesms, PID: 7988
11-01 22:31:37.868  7988  8071 E AndroidRuntime: java.lang.RuntimeException: An error occurred while executing doInBackground()
11-01 22:31:37.868  7988  8071 E AndroidRuntime:        at android.os.AsyncTask$4.done(AsyncTask.java:399)
11-01 22:31:37.868  7988  8071 E AndroidRuntime:        at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
11-01 22:31:37.868  7988  8071 E AndroidRuntime:        at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
11-01 22:31:37.868  7988  8071 E AndroidRuntime:        at java.util.concurrent.FutureTask.run(FutureTask.java:271)
11-01 22:31:37.868  7988  8071 E AndroidRuntime:        at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:289)
11-01 22:31:37.868  7988  8071 E AndroidRuntime:        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
11-01 22:31:37.868  7988  8071 E AndroidRuntime:        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
11-01 22:31:37.868  7988  8071 E AndroidRuntime:        at java.lang.Thread.run(Thread.java:919)
11-01 22:31:37.868  7988  8071 E AndroidRuntime: Caused by: net.sqlcipher.database.SQLiteException: no such table: recipient: , while compiling: SELECT _id FROM recipient WHERE phone = ?
11-01 22:31:37.868  7988  8071 E AndroidRuntime:        at net.sqlcipher.database.SQLiteCompiledSql.native_compile(Native Method)
11-01 22:31:37.868  7988  8071 E AndroidRuntime:        at net.sqlcipher.database.SQLiteCompiledSql.compile(SQLiteCompiledSql.java:3)
11-01 22:31:37.868  7988  8071 E AndroidRuntime:        at net.sqlcipher.database.SQLiteCompiledSql.<init>(SQLiteCompiledSql.java:12)
11-01 22:31:37.868  7988  8071 E AndroidRuntime:        at net.sqlcipher.database.SQLiteProgram.<init>(SQLiteProgram.java:18)
11-01 22:31:37.868  7988  8071 E AndroidRuntime:        at net.sqlcipher.database.SQLiteQuery.<init>(SQLiteQuery.java:1)
11-01 22:31:37.868  7988  8071 E AndroidRuntime:        at net.sqlcipher.database.SQLiteDirectCursorDriver.query(SQLiteDirectCursorDriver.java:8)
11-01 22:31:37.868  7988  8071 E AndroidRuntime:        at net.sqlcipher.database.SQLiteDatabase.rawQueryWithFactory(SQLiteDatabase.java:5)
11-01 22:31:37.868  7988  8071 E AndroidRuntime:        at net.sqlcipher.database.SQLiteDatabase.queryWithFactory(SQLiteDatabase.java:4)
11-01 22:31:37.868  7988  8071 E AndroidRuntime:        at net.sqlcipher.database.SQLiteDatabase.query(SQLiteDatabase.java:1)
11-01 22:31:37.868  7988  8071 E AndroidRuntime:        at net.sqlcipher.database.SQLiteDatabase.query(SQLiteDatabase.java:2)
11-01 22:31:37.868  7988  8071 E AndroidRuntime:        at org.thoughtcrime.securesms.database.RecipientDatabase.getByColumn(RecipientDatabase.java:3)
11-01 22:31:37.868  7988  8071 E AndroidRuntime:        at org.thoughtcrime.securesms.database.RecipientDatabase.getOrInsertByColumn(RecipientDatabase.java:2)
11-01 22:31:37.868  7988  8071 E AndroidRuntime:        at org.thoughtcrime.securesms.database.RecipientDatabase.getOrInsertFromE164(RecipientDatabase.java:1)
11-01 22:31:37.868  7988  8071 E AndroidRuntime:        at org.thoughtcrime.securesms.recipients.Recipient.external(Recipient.java:8)
11-01 22:31:37.868  7988  8071 E AndroidRuntime:        at org.thoughtcrime.securesms.backup.FullBackupImporter.processAvatar(FullBackupImporter.java:2)
11-01 22:31:37.868  7988  8071 E AndroidRuntime:        at org.thoughtcrime.securesms.backup.FullBackupImporter.importFile(FullBackupImporter.java:11)
11-01 22:31:37.868  7988  8071 E AndroidRuntime:        at org.thoughtcrime.securesms.registration.fragments.RestoreBackupFragment$2.doInBackground(RestoreBackupFragment.java:7)
11-01 22:31:37.868  7988  8071 E AndroidRuntime:        at org.thoughtcrime.securesms.registration.fragments.RestoreBackupFragment$2.doInBackground(RestoreBackupFragment.java:1)
11-01 22:31:37.868  7988  8071 E AndroidRuntime:        at android.os.AsyncTask$3.call(AsyncTask.java:378)
11-01 22:31:37.868  7988  8071 E AndroidRuntime:        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
11-01 22:31:37.868  7988  8071 E AndroidRuntime:        ... 4 more
```

**Comment by shaltayl** [comment]
Tested on the latest 4.49.14 same problem.

Nobody for help ?

**Comment by gitdode** [comment]
Same here on a Fairphone3 with Android 9:

`Caused by: net.sqlcipher.database.SQLiteException: no such table: recipient: , while compiling: SELECT _id FROM recipient WHERE phone = ?`

**Comment by greyson-signal** [comment]
Should be addressed in 4.49.15, rolling out now.

**Comment by shaltayl** [comment]
Hello, 

I have compiled the 4.49.15. 
The problem is resolved. 

Thank a lot. 
bye 



De: "Greyson Parrelli" <notifications@github.com> 
À: "signalapp/Signal-Android" <Signal-Android@noreply.github.com> 
Cc: "shal" <shal@free.fr>, "Author" <author@noreply.github.com> 
Envoyé: Lundi 4 Novembre 2019 18:08:06 
Objet: Re: [signalapp/Signal-Android] Restore backup, Crash on sql, On Smartphone or Emualtor (#9133) 



Should be addressed in 4.49.15, rolling out now.
